pkg/build/strategies/sti/sti.go

```diff
@@ -228,16 +228,27 @@ func (builder *STI) Build(config *api.Config) (*api.Result, error) {
 	} else {
 		glog.V(1).Infof("Running %q in %q", api.Assemble, config.Tag)
 	}
-	if err := builder.scripts.Execute(api.Assemble, config.AssembleUser, config); err != nil {
+	startTime := time.Now()
+	err := builder.scripts.Execute(api.Assemble, config.AssembleUser, config)
+	builder.result.BuildInfo.Steps = utilstatus.RecordStepInfo(builder.result.BuildInfo.Steps, api.StepAssembleBuildScripts, startTime)
+	if err != nil {
 		if err == errMissingRequirements {
 			glog.V(1).Info("Image is missing basic requirements (sh or tar), layered build will be performed")
 			return builder.layered.Build(config)
 		}
-		if _, ok := err.(s2ierr.ContainerError); ok {
-			builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
-				utilstatus.ReasonAssembleFailed,
-				utilstatus.ReasonMessageAssembleFailed,
-			)
+		if e, ok := err.(s2ierr.ContainerError); ok {
+			if !isMissingRequirements(e.Output) {
+				builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
+					utilstatus.ReasonAssembleFailed,
+					utilstatus.ReasonMessageAssembleFailed,
+				)
+				return builder.result, err
+			}
+			glog.V(1).Info("Image is missing basic requirements (sh or tar), layered build will be performed")
```

Comment:
So for a layered build you're going to do the following:
1) assemble runs and fails, you record an assemble duration (duration=1->1)
2) you start recording a LayeredBuild duration
3) you start recording a build docker image
4) you finish recording a build docker image (duration=3->4)
5) you start recording another assemble step
6) you finish recording that assemble step (will update the total duration of the assemble step from (1) to reflect the time from 1->6)
7) you finish recording the LayeredBuild step (duration=2->7)

I think we can agree that's going to be messy in terms of the data reported and how a user can interpret it...  some durations are being double counted, there is some overlap, etc.

I don't think we should ever nest step durations, and there should be no overlap between them..  
start stepA
stop stepA
start stepB
stop stepB

never:
startA
startB
stopB
stopA

and never:
startA
startB
stopA
stopB

In an ideal world, the total duration of all the steps recorded should == the total duration of the build.

In reality that won't be the case because we're not going to measure the times for every single step of the build process, so the total duration of the steps we record will be less than the total duration of the build.  but it definitely should not be more.

For the layered build case here, I think if you get an error executing assemble and conclude you're going to do a layered build instead, you should just throw away the original start time you noted, don't record the step time at all (the failed assemble invocation will be mostly a no-op).

And then within the layered build logic, record the DockerBuild duration and the Assemble duration.